### DISCUSSION TITLE
Airflow fetching too many tokens from Vault

### DISCUSSION THREAD

**Discussion by llamageddon83** [apache/airflow#30621] [General]
### Apache Airflow Provider(s)

hashicorp

### Versions of Apache Airflow Providers

3.3.0

### Apache Airflow version

2.5.3

### Operating System

linux

### Deployment

Official Apache Airflow Helm Chart

### What happened

We use Vault as the secrets backend for Airflow. And we have noticed quite an abnormal pattern where Airflow asks for 400 tokens every 10 minutes. This is when all the DAGs are disabled/paused. When the DAGs, totaling 300ish, are enabled and running, the number of tokens spikes to 40K. The TTL on the tokens is 15 minutes. Most of our DAGs finish within 10 minutes. 

We initially thought it was due to a typical pattern in the code related to how DAGs access secrets. But upon further troubleshooting, this happens during the auth, which is probably when the pod comes up. (We use Kubernetes Executor).

### What you think should happen instead

The token should be cached and Airflow should not be requesting a new token for every DAG execution. Even without any DAG executing, Airflow is asking for tokens.

### How to reproduce

Start Airflow with Kubernetes Executor and Vault as the secrets backend. Observe the number of tokens/leases in Vault. 

### Anything else

The following is after disabling all the DAGs in one of our lower environments. We can identify number of leases based on the read policy associated with Airflow.

![vault-leases](https://user-images.githubusercontent.com/108699169/231593538-2a96daad-8ced-47ab-9a55-c52882511c77.png)

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by potiuk** [comment]
Converted it into a discussion, as this is likely intended and not Airflow issue, but Airflow feature and result of violating the best practices.

Question @throwawayaccount11111 - Are you using Airflow Variables (https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/variables.html#variables) at the top-level of your DAGs? 

If this is the case - this is against the  "Best Practices" https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#top-level-python-code and you shoudl convert your code to use different approach to use dynamic /variable properties of dag generation - for example using BASH environment variables. Some of the approaches you can take are described here  https://airflow.apache.org/docs/apache-airflow/stable/howto/dynamic-dag-generation.html

Assuming that this is the reason:

Changing the behaviour entirely is not possible without re-architecting Airlfow internals, however we are discussing now a possible "temporary" or "tactical" solution and we have different voices on that and people who are supporting the "Tactical" solution with partial variable caching vs. waiting for architectural change.

More details in https://github.com/apache/airflow/pull/30259 and https://lists.apache.org/thread/ppbb87tohos9zs1yv6pf8b2zyq66dmdk

Since this is currently the problem for you, I would love to  hear your story to add a voice to the dicussion:

* Are your DAGs using Variables at top level?
* If so - why?
* Where did you get the idea from to use it?
* Can you use any of the alternatives (see above links) and if they are less feasible,  could you explain why?
* Why having more tokens allocated is a problem?
* Did you consider using Vault Caching agent instead (https://developer.hashicorp.com/vault/docs/agent/caching)  ? Is it a feasible solution?


Regardless from the decision about this (it seems difficult to get consensus) - having a voice of someone who hit such a problem would be really valuable.

**Comment by llamageddon83** [comment]
Hi @potiuk 

We had a pretty big scare with this issue. It is still ongoing. However, we lowered the TTL on the tokens from 24 hours to 1 hour and that has considerably improved the performance. The number of tokens being fetched is still a problem, though.

Thank you for the feedback and pointers. Please see the answers inline below:

- Are your DAGs using Variables at top level?

We do, but not at the top level. The other thing I would like to point out is that by the time DAG executes, Airflow has already performed auth with Vault. The auth happens during pod initialization. After that, it is just a read operation using the existing token/lease. We also performed the test by disabling all the DAGs and letting the tokens drain in Vault. That is how we were able to eliminate the implementation code as the root cause.

We also looked at the vault provider source code. And nothing jumped out. The provider code caches the credentials; checking before each `get_secret` operation to see if the Vault client is still authenticated, and recreating a new token if the authentication expires.

- If so - why?
For config management and ease of development. For example, a config in a config file `dev.yaml` (corresponding to the dev environment) could contain a config like this:

``` yaml
apples:
  bananas: 2
  oranges: replace_from_vault_SVC_ORANGES
```

In the code, the developer would call a helper method `get_config('apples')`, and the helper method would iterate over the keys under `apples` object, and if a value starts from `replace_from_vault`, the value is fetched from Vault using the `Variable.get_variable_from_secrets(key)` method. The returned value from `get_config('apples'), would be something like this:

```python
{'apples': {'bananas' :2, 'oranges': 'actual value fetched from vault'}}
```

- Where did you get the idea from to use it?
I do not recall. It has been a while since we actually started using Airflow and since then it has just become a pattern.

- Can you use any of the alternatives (see above links) and if they are less feasible, could you explain why?
Yes, most definitely. I will check with our engineering team.

- Why having more tokens allocated is a problem?
I am sure as you are aware, the reasoning behind Vault is that it is centralized across applications/systems. Too many tokens can cause disk space issues thereby potentially causing outages not just to Airflow but to anything that uses Vault which is literally everything. 

- Did you consider using Vault Caching agent instead (https://developer.hashicorp.com/vault/docs/agent/caching) ? Is it a feasible solution?
We did not. We do not do anything explicitly with Vault in the code. Airflow runs in AWS EKS. We manage deployments via helm charts. Everything in the DAG is using the code provided out of the box by Airflow.

**Reply by potiuk** [reply]
> We do, but not at the top level. 

I think you might be wrong. I would love to see snippets of your DAGS where variables are accessed. 

> We also looked at the vault provider source code. And nothing jumped out. The provider code caches the credentials; checking before each get_secret operation to see if the Vault client is still authenticated, and recreating a new token if the authentication expires.

This is not the "vault" code. Each DAG is parsed separately in a freshly forked process. This means that every time a DAG is parsed, the cache is empty because the process to parse the DAG has just been forked with empty cache.

**Comment by llamageddon83** [comment]
We decreased the TTL on tokens even further from 6 hours to 1 hour. Things have stabilized since.

**Reply by potiuk** [reply]
Thanks ! Good to know. We decided since not to implement a "temporary" solution for that but focus more on the "strategic" approach to improve it, let's see how it works out.

**Comment by infiniityr** [comment]
I'm concern that this is not considered as a bug...
When you start to have multiple DAGs, Airflow is making millions of requests per day and it is concerning...

I have 2 examples where I would be curious to know your advice.

First, in the configuration file.
In the documentation, we can use an external backend secrets (https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/security/secrets/secrets-backend/index.html#secrets-backend-configuration).
We felt it was a great solution to store our passwords in Vault so we used it in the `airflow.cfg`.
But, after some digging, Airflow makes 10 requests per second on each secret variable ! Why so ? It should only make 1 at the launch of the service...

`airflow-scheduler.service` :
```
[Service]
User=airflow
Environment='AIRFLOW_HOME=/path/to/airflow'
Environment='AIRFLOW__SECRETS__BACKEND=airflow.providers.hashicorp.secrets.vault.VaultBackend'
Environment='AIRFLOW__SECRETS__BACKEND_KWARGS={"connections_path": null, "variables_path": "variables", "config_path": "variables", "mount_point": "airflow", "url": "https://our.vault", "auth_type": "approle", "role_id": "role", "secret_id": "secret", "verify": "/path/to/certs"}'
```

`airflow.cfg` :
```
[database]
sql_alchemy_conn_secret = sql_alchemy_conn
```


Another example is in the code of a DAG.
Here I applied (maybe wrongly) your recommendations vut I have even more calls per minutes (around 100 calls / minutes).
In order to organize my code, I used classes.

The dag file :
```python
from facturation.task_facturator import TaskFacturator

DAG_VERSION = "1.1.17"
DAG_NAME = "our_dag"

with DAG(
    dag_id=f"{DAG_NAME}_v{DAG_VERSION}",
    schedule=None,
    default_args=DEFAULT_ARGS,
    catchup=False,
    max_active_runs=1,
) as dag:
    
    facturator = PythonOperator(
        task_id="facturator", python_callable=TaskFacturator().callable
    )
```

`task_facturator.py`:
```python
# imports
class TaskFacturator:
    def callable(self, task_instance):

        pg_hook = PostgresHook("database")

        line = PostgresHookHandler.run(
            pg_hook=pg_hook, request=self.request, handler=fetch_one_handler
        )
        return batch_id

    @property
    def request(self):
        date_facture = datetime.strptime(
            Variable.get("date"), "%Y-%m-%d"
        )
        return f"SELECT * WHERE date = '{date_facture.strftime('%Y-%m-%d')}';"
```


Do you have a way to explain why there are so many calls in Airflow ?
And also... Why do you generate SO MUCH tokens ? Airflow is not structure to have a singleton to optimize the use of this token ?

**Reply by potiuk** [reply]
I tihnk, if you want to express your concerns - I invite you to present a user opinion in this draft PR (which does not entirely remove the issue - but somewhat mitgates parts of it - not all ). https://github.com/apache/airflow/pull/30259. 

## Reason for the problem (and solution) 

The root cause is that you are not following good practices and you are using vault in top-level code of your DAGs. This is just wrong. And you should not do it: https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#top-level-python-code. Once you fix your code and move your `Variable.get()`  calls to within execute() methods of your tasks, things will be much smoother. In tjis case your request property of callable is executed every single time, every single DAG using it is parsed (all your DAGs are parsed with min 30 sec frequency by default). 

Just get rid of that call at top-level Parsing and you will be good.

## The longer explanation: 

Parsing each DAG and executing it, is done in a separte sub-process which can be done not only in separate sub-processes but also on separate machines. The same top-level parsing happens every 30 secods (or every [min_file_process_interval seconds](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#min-file-process-interval) interval). Changing that behaviour (and frequent parsing) is not an easy thing to change without re-architecting whole Airlfow architecture, for now the best way for you is to follow the best practices and do not call Vault at the Top-Level Python code (i.e. when the python DAG files of yours are parsed).

## Other solutions

Another solution if you want to stick to your top-level Python code vault usage, is to put Vault caching agent proxying your requests to vault. https://developer.hashicorp.com/vault/docs/agent/caching - this is solution suggested by Vault if you have multiple processes and nodes accessing Vault frequently - you install the caching agent and proxy your requests.

## Alternative options (for now rejected):

The PR in question https://github.com/apache/airflow/pull/30259 partially mitigates it in certain configurations, but likely you will still have bigger number of requests. Though in your case it could  actually help a lot, I think. 

However we have a very heated debate (aas you might see in this PR) whether we should implement it, because one of the main problems it has is that it encourages the behaviour that is not really good practice, it has side effects and does not solve the problem entirely. There are opinions that it is much better to explain people like you, who try to use Variables.get() level about the best practices and that you should just follow them and we should focus on future changes in the architecture  (which hereby I do as we agreed that at leaast for now we are not implementing this PR).

cc: @jedcunningham @vandonr-amz  @eladkal @uranusjr @john-jac @shubham22 @ashb and others who took part in the discussion. 

Maybe you have something to add ?

**Reply by infiniityr** [reply]
Thanks for the PR, I'll look into it in detail.

Concerning the problems indicated and the solution proposed, I am not sure to understand them well.
How is the `Variable.get()` used in top level ?
Isn't that exactly what is done in the 2nd "Good Example" (https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#airflow-variables) ?
The problem is in the import of classes?

Regarding the configuration file, so it is not recommended to use a backend secret inside ?
If so, shouldn't the `*_secret` variables be deprecated ?

**Reply by potiuk** [reply]
> Isn't that exactly what is done in the 2nd "Good Example" (https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#airflow-variables) ?
The problem is in the import of classes?

No. It is very different. It's about parsing Python code. Check it yourself. Raise an exception in your method just before the Variable.get() and parse your code (`python your_dag_file.py` -> exception will be raised (it will not be raised in the good example).


> Regarding the configuration file, so it is not recommended to use a backend secret inside ? If so, shouldn't the *_secret variables be deprecated ?

I am not sure what you mean about using backend secret inside. The `_secret` variables are declaring the secrets to read, nor readong the secrets, the secret is accessed in the Airflow code when the variables are used, so I do not know what you are referring to.

**Reply by vandonr-amz** [reply]
I can see how it can be confusing here because I'm confused myself... `callable` is passed as a function pointer, so I wouldn't expect the function body to be evaluated when just parsing the code ?

**Reply by potiuk** [reply]
But `self.request` is a property and it is used to calculate  a default value. 'self.request` is a property - and in order to calculate the default value you need to execute the method. 

This is why I suggested to raise exception inside the method. This is an easy way to check it.

Yes. Parsing Python code has some gotchas.

**Reply by infiniityr** [reply]
> I can see how it can be confusing here because I'm confused myself... `callable` is passed as a function pointer, so I wouldn't expect the function body to be evaluated when just parsing the code ?

It have the exact same confusion !
I tried some parsing with the snippet I wrote (addind logs and Exception) and the exception in the `request` property is never thrown !
As @vandonr-amz is saying, the `request` property is only use in the `callable` function which is never directly call.

Here is the exact code used :

```python
from datetime import datetime

from airflow import DAG
from airflow.operators.python import PythonOperator

class Test:
    def callable(self):
        print("In callable function")
        return self.request

    @property
    def request(self):
        raise Exception("In request property")
        return ""


DEFAULT_ARGS = {
    "start_date": datetime(2023, 1, 1)
}

with DAG(
    dag_id=f"test",
    schedule=None,
    default_args=DEFAULT_ARGS,
    catchup=False,
    max_active_runs=1,
) as dag:
    print("In DAG")
    test = PythonOperator(
        task_id="test_dag",
        python_callable=Test().callable,
    )

    test
```
-----------

I also insist on the configuration file.
As specify in the documentation, (https://airflow.apache.org/docs/apache-airflow/stable/howto/set-config.html), the Secret Backends can be use in the `airflow.cfg` file thought the `*_secret` variables.

In other words, we can externalize the configuration of the database into Vault.
Which is really really good ! But there are way too much requests.

In the example you provide, is it suppose to make 10 requests/minutes ?
```
[database]
sql_alchemy_conn_secret = sql_alchemy_conn
# You can also add a nested path
# example:
# sql_alchemy_conn_secret = database/sql_alchemy_conn
```
